tests/zfs-tests/include/kstat.shlib

```diff
@@ -0,0 +1,507 @@
+#
+# Output is <pool>/<objsetID>. objsetID is a decimal integer, > 0
+#
+
+####################
+# FreeBSD
+
+#
+# All kstats are accessed through sysctl. We model "groups" as interior nodes
+# in the stat tree, which are normally opaque. Because sysctl has no filtering
+# options, and requesting any node produces all nodes below it, we have to
+# always get the name and value, and then consider the output to understand
+# if we got a group or a single stat, and post-process accordingly.
+#
+# Scopes are mostly mapped directly to known locations in the tree, but there
+# are a handful of stats that are out of position, so we need to adjust.
+#
+
+#
+# _kstat_freebsd <scope> <object> <stat> <want_group>
+#
+function _kstat_freebsd
+{
+	typeset scope=$1
+	typeset obj=$2
+	typeset stat=$3
+	typeset -i want_group=$4
+
+	typeset oid=""
+	case "$scope" in
+	global)
+		oid="kstat.zfs.misc.$stat"
+		;;
+	pool)
+		# For reasons unknown, the "multihost", "txgs" and "reads"
+		# pool-specific kstats are directly under kstat.zfs.<pool>,
+		# rather than kstat.zfs.<pool>.misc like the other pool kstats.
+		# Adjust for that here.
+		case "$stat" in
+		multihost|txgs|reads)
+		    oid="kstat.zfs.$obj.$stat"
+		    ;;
+		*)
+		    oid="kstat.zfs.$obj.misc.$stat"
+		    ;;
+		esac
+		;;
+	dataset)
+		typeset pool=""
+		typeset -i objsetid=0
+		_split_pool_objsetid $obj pool objsetid
+		oid=$(printf 'kstat.zfs.%s.dataset.objset-0x%x.%s' \
+		    $pool $objsetid $stat)
+		;;
+	esac
+
+	# Calling sysctl on a "group" node will return everything under that
+	# node, so we have to inspect the first line to make sure we are
+	# getting back what we expect. For a single value, the key will have
+	# the name we requested, while for a group, the key will not have the
+	# name (group nodes are "opaque", not returned by sysctl by default.
+
+	if [[ $want_group == 0 ]] ; then
+		sysctl -e "$oid" | awk -v oid="$oid" -v oidre="^$oid=" '
+			NR == 1 && $0 !~ oidre { exit 1 }
+			NR == 1 { print substr($0, length(oid)+2) ; next }
+			{ print }
+		'
+	else
+		sysctl -e "$oid" | awk -v oid="$oid" -v oidre="^$oid=" '
+			NR == 1 && $0 ~ oidre { exit 2 }
+			{
+			    sub("^" oid "\.", "")
+			    sub("=", " ")
+			    print
+			}
+		'
+	fi
+
+	typeset -i err=$?
+	case $err in
+		0) return ;;
+		1) log_fail "kstat: can't get value for group kstat: $oid" ;;
+		2) log_fail "kstat: not a group kstat: $oid" ;;
+	esac
+
+	log_fail "kstat: unknown error: $oid"
+}
+
+function _resolve_dsname_freebsd
+{
+	# we're searching for:
+	#
+	# kstat.zfs.shed.dataset.objset-0x8087.dataset_name: shed/poudriere
+	#
+	# We split on '.', then return the pool name from field 3, and the
+	# hex objsetid from field 5.
+	#
+	# We convert hex to decimal in the shell because there isn't a _simple_
+	# portable way to do it in awk and this code is already too intense to
+	# do it a complicated way.
+	typeset dsname=$1
+	sysctl -e kstat.zfs | \
```

Comment:
Should not be a problem, but I wonder if passing the pool name here to reduce the output would be more efficient.

Reply:
Yeah, good idea. Produces a slightly smaller mess :)